crates/stages/stages/src/stages/merkle_changesets.rs

```diff
@@ -0,0 +1,374 @@
+                },
+            )?;
+
+        let block = provider
+            .header_by_number(block_number)?
+            .ok_or_else(|| ProviderError::HeaderNotFound(block_number.into()))?;
+
+        let header = SealedHeader::seal_slow(block);
+
+        let (got, expected) = (root, header.state_root());
+        if got != expected {
+            error!(
+                target: "sync::stages::merkle_changesets",
+                ?block_number,
+                ?got,
+                ?expected,
+                "Failed to verify block state root! {INVALID_STATE_ROOT_ERROR_MESSAGE}",
+            );
+            return Err(StageError::Block {
+                error: BlockErrorKind::Validation(ConsensusError::BodyStateRootDiff(
+                    GotExpected { got, expected }.into(),
+                )),
+                block: Box::new(header.block_with_parent()),
+            })
+        }
+
+        Ok(trie_updates)
+    }
+
+    fn populate_range<Provider>(
+        provider: &Provider,
+        target_range: Range<BlockNumber>,
+    ) -> Result<(), StageError>
+    where
+        Provider: StageCheckpointReader
+            + TrieWriter
+            + DBProvider
+            + HeaderProvider
+            + ChainStateBlockReader,
+    {
+        let target_start = target_range.start;
+        let target_end = target_range.end;
+        debug!(
+            target: "sync::stages::merkle_changesets",
+            ?target_range,
+            "Starting trie changeset computation",
+        );
+
+        // We need to distinguish a full revert and a per-block revert. A full revert reverts
+        // changes starting at db tip all the way to a block. A per-block revert only reverts
+        // a block's changes.
+        //
+        // We need to calculate the full HashedPostState reverts for every block in the target
+        // range. The full HashedPostState revert for block N can be calculated as:
+        //
+        //
+        // ```
+        // // where `extend` overwrites any shared keys
+        // state_revert(N) = state_revert(N + 1).extend(per_block_state_revert(N))
+        // ```
+        //
+        // We need per-block reverts to calculate the prefix set for each individual block. By using
+        // the per-block reverts to calculate full reverts on-the-fly we can save a bunch of memory.
+        debug!(
+            target: "sync::stages::merkle_changesets",
+            ?target_range,
+            "Computing per-block state reverts",
+        );
+        let mut per_block_state_reverts = Vec::new();
+        for block_number in target_range.clone() {
+            per_block_state_reverts.push(HashedPostState::from_reverts::<KeccakKeyHasher>(
+                provider.tx_ref(),
+                block_number..=block_number,
+            )?);
+        }
+
+        let per_block_state_revert = |block_number| -> &HashedPostState {
+            &per_block_state_reverts[(block_number - target_start) as usize]
+        };
```

Comment:
might be a nit from my side, but feels like this closure syntax is not immediately intuitive? 

i.e retries state revert data for given block from array 

maybe a docstring / bringing this out as a pure function could help?